src/psyclone/dynamo0p3.py

```diff
@@ -3687,10 +3716,178 @@ def gen_code(self, parent):
+
+    def _compute_halo_depth(self):
+        '''Call the required method in the corresponding halo exchange end
+        object. This is done as the field in halo exchange start is
+        only read and the dependence analysis beneath this call
+        requires the field to be modified.
+
+        :return: Return the halo exchange depth as a Fortran string
+        :rtype: str
+
+        '''
+        return self._get_hex_end()._compute_halo_depth()
+
+    def required(self):
+        '''Call the required method in the corresponding halo exchange end
+        object. This is done as the field in halo exchange start is
+        only read and the dependence analysis beneath this call
+        requires the field to be modified.
+
+        :return: Returns (x, y) where x specifies whether this halo \
+        exchange is (or might be) required - True, or is not required \
+        - False. If the first argument is True then the second \
+        argument specifies whether we definitely know that we need the \
+        HaloExchange - True, or are not sure - False.
+        :rtype: (bool, bool)
+
+        '''
+        return self._get_hex_end().required()
+
+    def _get_hex_end(self):
+        '''An internal helper routine for this class which finds the halo
+        exchange end object corresponding to this halo exchange start
+        object or raises an exception if one is not found.
+
+        :return: The corresponding halo exchange end object
+        :rtype: :py:class:`psyclone.dynamo0p3.DynHaloExchangeEnd`
+        :raises GenerationError: If no matching HaloExchangeEnd is \
+        found, or if the first matching haloexchange that is found is \
+        not a HaloExchangeEnd
+
+        '''
+        # Look at all nodes following this one in schedule order
+        # (which is PSyIRe node order)
+        for node in self.following():
+            if self.sameParent(node) and isinstance(node, DynHaloExchange):
+                # Found a following `haloexchange`,
+                # `haloexchangestart` or `haloexchangeend` PSyIRe node
+                # that is at the same calling hierarchy level as this
+                # haloexchangestart
+
+                # TBD: support matching fields with the same vector
+                # index and skipping ones with different indices. This
+                # can be added simply once PR #225 is on master as
+                # that adds the concept of matching overlapping
+                # accesses. Note, we still get correct results if no
+                # move transformations are applied.
+                if node.field.name == self.field.name:
+                    # the names match so it should be the associated
+                    # haloexchangeend (unless it has been moved by a
+                    # transformation).
+                    if isinstance(node, DynHaloExchangeEnd):
+                        return node
+                    else:
+                        raise GenerationError(
```

Comment:
This line is not covered (according to CodeCov) and nor is the `raise GenerationError` at line 3849 below.

Reply:
I think my excuse for this was that I was waiting for the vector dependence logic to be fixed first. This has now been fixed and is on master and I've added the missing tests.